src/ln/functional_tests.rs

```diff
@@ -5352,3 +5352,261 @@ fn test_update_fulfill_htlc_bolt2_after_malformed_htlc_message_must_forward_upda
+
+	if let Err(err) = nodes[1].node.handle_commitment_signed(&nodes[0].node.get_our_node_id(), &invalid_commitment) {
+		check_added_monitors!(nodes[1], 0);
+		if let msgs::ErrorAction::SendErrorMessage { msg } = err.action.unwrap() {
+			assert_eq!(msg.channel_id, channel_id);
+			assert_eq!(msg.data, "Invalid commitment tx signature from peer");
+		} else {
+			panic!("Unexpected ErrorAction");
+		}
+	} else {
+		panic!("Expected channel failure but that did not happen");
+	}
+
+	assert!(nodes[1].node.list_channels().is_empty());
+	check_closed_broadcast!(nodes[1]);
+}
+
+#[test]
+fn test_commitment_signed_bolt2_receiving_node_pending_updates_applied_num_htlc_output_incorrect_fail_channel() {
+	//BOLT 2 Requirement: A receiving node:
+	//    * once all pending updates are applied: if num_htlcs is not equal to the number of HTLC outputs in the local commitment transaction: MUST fail the channel.
+
+	let mut nodes = create_network(2);
+	let (_, _, channel_id, _) = create_announced_chan_between_nodes(&nodes, 0, 1);
+
+	let route = nodes[0].router.get_route(&nodes[1].node.get_our_node_id(), None, &[], 1000000, TEST_FINAL_CLTV).unwrap();
+	let (_, our_payment_hash) = get_payment_preimage_hash!(nodes[0]);
+
+	nodes[0].node.send_payment(route, our_payment_hash).unwrap();
+	check_added_monitors!(nodes[0], 1);
+
+	let updates = get_htlc_update_msgs!(nodes[0], nodes[1].node.get_our_node_id());
+
+	nodes[1].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &updates.update_add_htlcs[0]).unwrap();
+
+	let mut invalid_commitment = updates.commitment_signed.clone();
+	invalid_commitment.htlc_signatures.pop();
+
+	if let Err(err) = nodes[1].node.handle_commitment_signed(&nodes[0].node.get_our_node_id(), &invalid_commitment) {
+		check_added_monitors!(nodes[1], 0);
+		if let msgs::ErrorAction::SendErrorMessage { msg } = err.action.unwrap() {
+			assert_eq!(msg.channel_id, channel_id);
+			assert_eq!(msg.data, "Got wrong number of HTLC signatures from remote");
+		} else {
+			panic!("Unexpected ErrorAction");
+		}
+	} else {
+		panic!("Expected channel failure but that did not happen");
+	}
+
+	assert!(nodes[1].node.list_channels().is_empty());
+	check_closed_broadcast!(nodes[1]);
+}
+
+#[test]
+fn test_commitment_signed_bolt2_receiving_node_any_htlc_signature_invalid_fail_channel() {
+	//BOLT 2 Requirement: A receiving node:
+	//    * if any htlc_signature is not valid for the corresponding HTLC transaction: MUST fail the channel.
+
+	use secp256k1::ffi::Signature as FFISignature;
+	use secp256k1::Signature;
+
+	let mut nodes = create_network(2);
+	let (_, _, channel_id, _) = create_announced_chan_between_nodes(&nodes, 0, 1);
+
+	let route_0 = nodes[0].router.get_route(&nodes[1].node.get_our_node_id(), None, &[], 1000000, TEST_FINAL_CLTV).unwrap();
+
+	let (_, our_payment_hash) = get_payment_preimage_hash!(nodes[0]);
+	nodes[0].node.send_payment(route_0, our_payment_hash).unwrap();
```

Comment:
You may send more than one htlc at once, at least to test if receiver handle ordering well